### PR TITLE
CORDA-902 - AMQP Setter Construction when empty / no constructor

### PR DESCRIPTION
**Pull request by fenryka**

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 1]
```diff
@@ -103,15 +121,45 @@ private fun <T : Any> propertiesForSerializationFromConstructor(kotlinConstructo
...
-    return rc
+
+    return ConstructorDestructorMethods(rc, emptyList())
+}
+
+/**
+ * If we determine a class has a constructor that takes no parameters then check for pairs of getters / setters
+ * and use those
+ */
+private fun propertiesForSerializationFromGetters(
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: FromSetters

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 2]
```diff
@@ -103,15 +121,45 @@ private fun <T : Any> propertiesForSerializationFromConstructor(kotlinConstructo
...
+    val getters : MutableList<PropertySerializer> = ArrayList(properties.size)
+    val setters : MutableList<Method?> = ArrayList(properties.size)
+
+    properties.forEach { property ->
+        val getter: Method? = property.value.readMethod
+        val setter: Method? = property.value.writeMethod
+
+        if (getter == null || setter == null) return@forEach
+
+        // NOTE: For now we're ignoring type mismatches. I.e. where the setter or getter type differs
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: Let's word this carefully.   Maybe don't say ignore.  Excluded perhaps?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 3]
```diff
@@ -60,22 +60,58 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
+                readObjectBuildViaSetters(obj, schemas, input)
+            }
         } else throw NotSerializableException("Body of described type is unexpected $obj")
     }
 
+    private fun readObjectBuildViaConstructor(
+            obj: List<*>,
+            schemas: SerializationSchemas,
+            input: DeserializationInput) : Any = ifThrowsAppend({ clazz.typeName }){
+        logger.debug { "Calling construction based construction" }
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: If there is a trace level, let's drop it down to that level.  And the others.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 1 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 2 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 4]
```diff
@@ -60,22 +60,58 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
+                readObjectBuildViaSetters(obj, schemas, input)
+            }
         } else throw NotSerializableException("Body of described type is unexpected $obj")
     }
 
+    private fun readObjectBuildViaConstructor(
+            obj: List<*>,
+            schemas: SerializationSchemas,
+            input: DeserializationInput) : Any = ifThrowsAppend({ clazz.typeName }){
+        logger.trace ("Calling construction based construction for ${clazz.typeName}")
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: Can you use the lazy version please?  `logger.trace { "" }`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 5]
```diff
@@ -60,22 +60,58 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
+
+        return construct(obj.zip(propertySerializers.getters).map { it.second.readProperty(it.first, schemas, input) })
+    }
+
+    private fun readObjectBuildViaSetters(
+            obj: List<*>,
+            schemas: SerializationSchemas,
+            input: DeserializationInput) : Any = ifThrowsAppend({ clazz.typeName }){
+        logger.trace ("Calling setter based construction for ${clazz.typeName}")
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: Make lazy also.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt [hunk 6]
```diff
@@ -60,22 +60,58 @@ open class ObjectSerializer(val clazz: Type, factory: SerializerFactory) : AMQPS
...
+        return propertySerializers.getters.map {
+            Field(it.name, it.type, it.requires, it.default, null, it.mandatory, false)
+        }
     }
 
     private fun generateProvides(): List<String> = interfaces.map { nameForType(it) }
 
     fun construct(properties: List<Any?>): Any {
-
         logger.debug { "Calling constructor: '$javaConstructor' with properties '$properties'" }
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: I know this was already there, but let's make this trace too.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 4 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: yes!

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 5 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt (hunk 6 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/ObjectSerializer.kt]: also done

**Comment by rick-r3** [comment]
Approved on the understanding a docs PR follows

**Pull request merged**